### PR TITLE
Coordinates Module

### PR DESCRIPTION
**Pull request by MaxNoe**
This is work in progress on a new coordinate trafo module.

I open this PR to get early Feedback from y'all.

This is addressing a long standing issue: #77

### CODE REVIEW

**Comment by MaxNoe** [comment]
The functionality is now complete. Please have a look.

As soon as we agree on the basic structure, I will start refactoring the Processors to use the new coordinates module.

File: src/main/java/fact/coordinates/CameraCoordinate.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+import org.apache.commons.math3.geometry.euclidean.threed.Rotation;
+import org.apache.commons.math3.geometry.euclidean.threed.Vector3D;
+
+import java.io.Serializable;
+
+/**
+ * Created by maxnoe on 22.05.17.
+ */
+public class CameraCoordinate implements Serializable {
```
Comment (mackaiver) [inline src/main/java/fact/coordinates/CameraCoordinate.java]: Missing javadoc

File: src/main/java/fact/coordinates/CoordinateUtils.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    private static final ZonedDateTime gstReferenceDateTime = ZonedDateTime.of(2000, 1, 1, 12, 0, 0, 0, ZoneId.of("UTC"));
+
+
+    /**
+     * Convert a DateTime object to greenwhich sidereal time according to
+     * https://en.wikipedia.org/wiki/Sidereal_time#Definition
+     * @param datetime
+     * @return gst in radians
+     */
+    public static double datetimeToGST(ZonedDateTime datetime){
```
Comment (mackaiver) [inline src/main/java/fact/coordinates/CoordinateUtils.java]: If I see this correctly, this only ever gets used in objects of type CelestialCoordinate. In that cas eit makes sense to move that function to the interface as has been done with the greatCircleDistance method.

File: src/main/java/fact/coordinates/CelestialCoordinate.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
+package fact.coordinates;
+
+import java.io.Serializable;
+
+/**
+ * Created by maxnoe on 23.05.17.
+ */
+public interface CelestialCoordinate extends Serializable {
```
Comment (mackaiver) [inline src/main/java/fact/coordinates/CelestialCoordinate.java]: missing javadoc

File: src/main/java/fact/coordinates/CameraCoordinate.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    public double getXMM() {
+        return xMM;
+    }
+
+    public double getYMM() {
+        return yMM;
+    }
+
+
+    public HorizontalCoordinate toHorizontal(HorizontalCoordinate pointingPosition, double focalLength){
```
Comment (mackaiver) [inline src/main/java/fact/coordinates/CameraCoordinate.java]: Very important public method. Should have extensive java docs.

File: src/main/java/fact/coordinates/EarthLocation.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
+package fact.coordinates;
+
+/**
+ * Created by maxnoe on 22.05.17.
+ */
+public class EarthLocation {
```
Comment (mackaiver) [inline src/main/java/fact/coordinates/EarthLocation.java]: missing javadoc

File: src/main/java/fact/coordinates/EarthLocation.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    final private double longitudeRad;
+    final private double altitudeMeter;
+
+    public EarthLocation(double latitudeRad, double longitudeRad, double altitudeMeter) {
+        this.latitudeRad = latitudeRad;
+        this.longitudeRad = longitudeRad;
+        this.altitudeMeter = altitudeMeter;
+    }
+
+    public static EarthLocation fromDegrees(double latitudeDeg, double longitudeDeg, double altitudeMeter) {
```
Comment (mackaiver) [inline src/main/java/fact/coordinates/EarthLocation.java]: I think it makes more sense to provide another `fromRads` constructor and make the cosntructor private. Seems to be more consistent.

File: src/main/java/fact/coordinates/EquatorialCoordinate.java [hunk 7]
```diff
@@ -0,0 +1,103 @@
+package fact.coordinates;
+
+import java.time.ZonedDateTime;
+
+/**
+ * Created by maxnoe on 22.05.17.
+ */
+public class EquatorialCoordinate implements CelestialCoordinate {
```
Comment (mackaiver) [inline src/main/java/fact/coordinates/EquatorialCoordinate.java]: missing javadoc

File: src/main/java/fact/coordinates/EquatorialCoordinate.java [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+
+    private final double rightAscensionRad;
+    private final double declinationRad;
+
+    public EquatorialCoordinate(double rightAscensionRad, double declinationRad) {
+        this.rightAscensionRad = rightAscensionRad;
+        this.declinationRad = declinationRad;
+    }
+
+    public static EquatorialCoordinate fromDegrees(double rightAscensionDeg, double declinationDeg) {
```
Comment (mackaiver) [inline src/main/java/fact/coordinates/EquatorialCoordinate.java]: Add `fromRad` method. for consistency.

File: src/main/java/fact/coordinates/EquatorialCoordinate.java [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+        }
+
+        return new HorizontalCoordinate(Math.PI / 2.0 - altitude, azimuth);
+    }
+
+    public String toString(){
+        return String.format("HorizontalCoordinate(ra=%.4f ha, dec=%.4f°)", this.getRightAscensionHA(), this.getDeclinationDeg());
+    }
+
+    public double greatCircleDistance(EquatorialCoordinate other) {
```
Comment (mackaiver) [inline src/main/java/fact/coordinates/EquatorialCoordinate.java]: missing javadoc

File: src/main/java/fact/coordinates/HorizontalCoordinate.java [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+
+        double declination = Math.asin(Math.sin(lat) * Math.sin(alt) - Math.cos(lat) * Math.cos(alt) * Math.cos(azimuthSouth));
+
+        double ra = gst + lon - hourAngle;
+
+        return new EquatorialCoordinate(ra, declination);
+    }
+
+    /**
+     * Returns position of the source in the camera from the given pointing position
```
Comment (mackaiver) [inline src/main/java/fact/coordinates/HorizontalCoordinate.java]: I find this confusing. Does it return the position of a source in the camera or does it transform the horizontal coordinate called `pointingPosition` to camera coordinates?

File: src/main/java/fact/coordinates/HorizontalCoordinate.java [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+    public double greatCircleDistance(HorizontalCoordinate other) {
+        return CelestialCoordinate.greatCircleDistance(
+                this.getAltitudeRad(), this.getAzimuthRad(),
+                other.getAltitudeRad(), other.getAzimuthRad()
+        );
+    }
+
+
+    /**
+     * Implementation of the formulas from
```
Comment (mackaiver) [inline src/main/java/fact/coordinates/HorizontalCoordinate.java]: One sentence to describe what this does wouldn't hurt.

File: src/main/java/fact/coordinates/EarthLocation.java (hunk 5 above)
Comment (mackaiver) [inline src/main/java/fact/coordinates/EarthLocation.java]: Provide convenience constructors for La Palma, Paranal and whatnot?

**Comment by MaxNoe** [comment]
@mackaiver I think I implemented all your suggestions, could you have a look?

**Comment by MaxNoe** [comment]
As nice as this is, we are probably better of using pyfact and the classifier tools for this and let FACT-Tools only do the Raw Data Processing. What do you think?

**Comment by dneise** [comment]
I am not taking part in fact tools development at all, but this sounds so good, I wanted to express my support:

Awesome

**Comment by MaxNoe** [comment]
I Implemented precession correction, and now the deviation to astropy is much smaller than a pixel (the mean is about 0.004° and the max is 0.008° )

**Comment by MaxNoe** [comment]
@fact-project/fact-tools-core-dev I think this is ready to merge now.

Please review (again)!

* Complete set of coordinate transformations (To a precision < 0.02°) inside of astropy
* Split up Theta and Disp processor
* Generally refactor `SourcePosition` Processor

File: src/main/java/fact/coordinates/EquatorialCoordinate.java [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+        double rightAscensionRad = rightAscensionHA / 12.0 * Math.PI;
+        return new EquatorialCoordinate(rightAscensionRad, declinationRad);
+    }
+
+    public static EquatorialCoordinate fromHourAngleAndDegrees(double rightAscensionHA, double declinationDeg) {
+        double rightAscensionRad = rightAscensionHA / 12.0 * Math.PI;
+        return new EquatorialCoordinate(rightAscensionRad, Math.toRadians(declinationDeg));
+    }
+
+    static double[] calculatePrecessionCorrection(double rightAscensionRad, double declinationRad, ZonedDateTime observationTime) {
```
Comment (mackaiver) [inline src/main/java/fact/coordinates/EquatorialCoordinate.java]: can we add a comment here citing the astronomical almanach. this is whre this comes from right?

File: src/main/java/fact/features/source/ReconstructSourcePositionDisp.java [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+        // orientated towards the suspected source position
+        double sign = - Math.signum(cosDeltaAlpha) * Math.signum(m3l * Math.signum(cosDeltaAlpha) - signM3lConstant);
+
+        double x = cogx + disp * Math.cos(delta) * sign;
+        double y = cogy + disp * Math.sin(delta) * sign;
+
+        return new CameraCoordinate(x, y);
+    }
+
+    public void setDispKey(String dispKey) {
```
Comment (mackaiver) [inline src/main/java/fact/features/source/ReconstructSourcePositionDisp.java]: I dont thionk we need getter and setters here

File: src/main/java/fact/coordinates/EquatorialCoordinate.java (hunk 12 above)
Comment (MaxNoe) [reply src/main/java/fact/coordinates/EquatorialCoordinate.java]: I will add the reference.

File: src/main/java/fact/features/source/ReconstructSourcePositionDisp.java (hunk 13 above)
Comment (MaxNoe) [reply src/main/java/fact/features/source/ReconstructSourcePositionDisp.java]: Yup.

**Comment by MaxNoe** [comment]
@mackaiver I implemented your comments

**Comment by MaxNoe** [comment]
I added two python scripts and a java executable to compare FACT-Tools with astropy, here is the result:

facttools_trafos_hist
facttools_trafos_time

So the maximal error on Crab in 50 years is 30 arcseconds.

**Comment by MaxNoe** [comment]
theta2-coordinates

**Pull request merged**